クラス（コマンドラインメモアプリ作成）

1mori opened the pull request:
## 関連リンク
課題ページ：https://bootcamp.fjord.jp/practices/208

## 最終更新
2024/06/20

## やりたいこと
- 各メソッドのクラス化

## なぜなのか
- 共通している処理（非同期化、クエリの実行）をまとめたいため
- クラスの課題であるため

## やったこと
- db_utils.jsで定義していたrunPromise, allPromise, closePromiseを`DbPromise`クラスにまとめ、run, all, closeに改名
- 標準入力のパッケージ`readline`での処理を`ReadlineInterface`クラスとして定義 
  - `createInterface`で使用している処理は`on`のみであり、Callback APIとPromise APIで書き方が変わらないのでCallback APIを採用
- クエリ文を実行するデータベース系の処理を`MemoDataBase`にまとめる
  - ensureTableExists, all, insert, delete, closeの5つを定義、MemoAppからクエリ系の処理を切り離した

- db_utils.jsの削除　内部で実行していたメソッドを`MemoDataBase`に
- 初回実行時にテーブルが存在しない場合、テーブル作成クエリを実行 (ensureTableExists)
- エラー処理の修正

## 見てほしいところ
- 全体の合理性の確認

1mori committed:
add memo.js

1mori committed:
必要なnpmパッケージを追加

1mori committed:
標準入力を受け付ける処理を追加

1mori committed:
シェバン処理を追加

1mori committed:
複数行のコメントの入力に対応

1mori committed:
add: ./memo.sqlite3ファイルを読み込む変数dbを定義

1mori committed:
add: runPromise, allPromiseを追加

1mori committed:
change: 標準入力される文字列に対し改行コードを入れるようにし、一つの文字列として扱えるように変更

1mori committed:
add: closePromiseを追加

1mori committed:
change: 非同期処理のエラー処理を追加

1mori committed:
mode: 非同期処理で必要なrunPromise, allPromise, closePromiseをdb_utils.jsに移動

1mori committed:
add: クラスMemoAppを定義

1mori committed:
定義の必要のあるクラス内関数add, list, read, deleteを定義

1mori committed:
define: メモの追加メソッドを作成

1mori committed:
fix: 閉じ括弧の追加

1mori committed:
define: メモ一覧を表示するメソッドの追加

1mori committed:
define: データベースを閉じるメソッドcloseDatabaseを追加

1mori committed:
change: rlの定義位置を変更

1mori committed:
define: runメソッドを定義、オプションを読んで実行したい項目のメソッドを実行する

1mori committed:
restore: コンストラクタ内でreadlineを宣言

1mori committed:
listメソッドで標準入力が終了するように.close()を宣言

1mori committed:
npm install @inquirer/prompts

1mori committed:
npm uninstall @inquirer/prompts, npm install --save inquirer

1mori committed:
import inquirer from inquirer

1mori committed:
add: readメソッドを実装

1mori committed:
add: deleteメソッドを実装

1mori committed:
change: エラーコンソール文の修正

1mori committed:
update: クラス内メソッドをプライベートメソッドに変更

1mori committed:
change: this.argv -> this.option

1mori committed:
change: this.rl -> this.userInput

1mori committed:
change: パッケージのインストール箇所を変更

1mori committed:
fix: lintが最終行の改行をわざわざ消してくれるので改行を追加

cafedomancer commented on 04.class/memo.sqlite3:
SQLite3 のデータベースファイルは git 管理しないでください。データを変更するたびに git 上で差分が生じてしまうからです。

1mori replied:
`.gitignore`に`.sqlite3`ファイルを含めないよう設定しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env node
+
+import * as readline from "node:readline/promises";
+import { stdin as input, stdout as output } from "node:process";
```
確認なのですが、`stdin` と `stdout` に対して別名を定義しているのはなぜですか？ `stdin` と `stdout` は単純な input や output とは異なる特別なものなので、`input` や `output` という単純な名前にしてしまうとそのような意味合いが失われてしまう気がします。

1mori replied:
公式ドキュメントに倣った書き方をしました。

https://nodejs.org/api/readline.html
> import { stdin as input, stdout as output } from 'node:process';

この書き方であれば省略記法が使えますが、意味は分かりにくいので修正します。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env node
+
+import * as readline from "node:readline/promises";
```
確認なのですが、default import するのではなく namespace import しているのはなぜですか？

1mori replied:
Node.jsの公式ドキュメントに倣った書き方をしました。

https://nodejs.org/api/readline.html
> import * as readline from 'node:readline/promises';

1mori replied:
下のコメントの段階で修正したので追記になりますが、callbackAPIのみ使用しているため
`import { createInterface } from "node:readline";`
に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+    }
+  }
+
+  async #closeDatabase() {
+    try {
+      await closePromise(this.db);
+    } catch (err) {
+      console.error(`Close error: ${err}`);
+    }
+  }
```
private メソッドであるとはいえ、これは `MemoApp` という名前のクラスに属するべき処理ではないと思います。データベースにしか関係のない処理で、メモアプリとは直接的な関係はないからです。

1mori replied:
クラスの外で定義し直しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+  #add() {
+    let inputText = "";
+
+    this.userInput.on("line", (line) => {
+      inputText += line + "\n";
+    });
+    this.userInput.on("close", async () => {
+      try {
+        await runPromise(this.db, "INSERT INTO memo (text) VALUES (?)", [
+          inputText,
+        ]);
+      } catch (err) {
+        console.error(`Insert error: ${err}`);
+      } finally {
+        await this.#closeDatabase();
+      }
+    });
```
`node:readline/promises` が import されているのに Promises API ではなく Callback API が使われているような...。

1mori replied:
今回はcallback APIのみ使用しているので、

`import { createInterface } from "node:readline";`
に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+        await runPromise(this.db, "INSERT INTO memo (text) VALUES (?)", [
+          inputText,
+        ]);
+      } catch (err) {
+        console.error(`Insert error: ${err}`);
+      } finally {
```
非同期処理のプラクティスでも触れていますが。例外を握り潰してはいけません。

1mori replied:
指摘いただいたにも関わらず修正しておらずすみません。
該当箇所は必要なエラーのみ捕捉するように変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+    let inputText = "";
+
+    this.userInput.on("line", (line) => {
+      inputText += line + "\n";
```
`line + "\n"` の部分に関して、文字列結合ではなくテンプレート文字列を使いましょう。

1mori replied:
修正しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+    this.userInput.on("line", (line) => {
+      inputText += line + "\n";
+    });
+    this.userInput.on("close", async () => {
+      try {
+        await runPromise(this.db, "INSERT INTO memo (text) VALUES (?)", [
+          inputText,
+        ]);
+      } catch (err) {
+        console.error(`Insert error: ${err}`);
+      } finally {
+        await this.#closeDatabase();
+      }
+    });
```
この実装だと Ctrl-C で入力を中断した場合でもメモが保存されてしまいます。

1mori replied:
`finally`から外しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+class MemoApp {
+  constructor() {
+    this.option = minimist(process.argv.slice(2));
+    this.userInput = readline.createInterface({ input, output });
```
`userInput` という名前だと「ユーザーが入力したもの (入力し終わっているもの) 」に見えるのですが、格納されているものは違うものなので名前を見直したほうがよさそうです。

1mori replied:
ドキュメントで用いられている`rl`はreadlineの挙動を知らない方に伝わりにくいかなと考え、上記の`userInput`にしました。
指摘を踏まえ、`readline.createInterface`を格納していることから`readlineInterface`という名前に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+        ]);
+      } catch (err) {
+        console.error(`Insert error: ${err}`);
+      } finally {
+        await this.#closeDatabase();
+      }
```
各メソッドでコネクションを切断するような実装にしていると、メソッドが増えたときに対応漏れが発生しそうです。共通化したほうがよいと思います。

1mori replied:
`run`メソッド内で共通化しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+    try {
+      const rows = await allPromise(this.db, "SELECT text FROM memo");
+      rows.forEach((row) => {
+        console.log(row["text"].split("\n")[0]);
```
`row["text"]` の部分に関して、ブラ-ケット記法ではなくドット記法を使ってください。ドット記法で参照できないようなプロパティ名ではないからです。

1mori replied:
修正しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+
+  async #list() {
+    try {
+      const rows = await allPromise(this.db, "SELECT text FROM memo");
```
`rows` だと抽象的すぎます。具体的な名前を付けましょう。

1mori replied:
`memoRows`に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  async #read() {
+    try {
+      const rows = await allPromise(this.db, "SELECT id, text FROM memo");
+
+      const choices = rows.map((row) => ({
+        name: row.text.split("\n")[0],
+        value: row.id,
+      }));
+
+      const answers = await inquirer.prompt([
+        {
+          name: "memoToRead",
+          type: "list",
+          message: "Choose a memo you want to read:",
+          choices: choices,
+        },
+      ]);
+
+      const memoIdToRead = answers.memoToRead;
+      const selectedMemo = rows.find((row) => row.id === memoIdToRead);
+      if (selectedMemo) {
+        console.log("\n" + selectedMemo.text);
+      }
+    } catch (err) {
```
try で囲む範囲が広すぎます。例外処理したい箇所がどこなのかが分かりづらいです。

1mori replied:
囲む範囲を限定しました。
その際、範囲を限定し`delete`メソッドと処理が重複していたので、`getMemoRows`メソッドを作成し、処理を共通化させました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+          name: "memoToRead",
+          type: "list",
+          message: "Choose a memo you want to read:",
+          choices: choices,
```
プロパティ名と値の変数が同じ時は省略記法を使ってください。

1mori replied:
変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+        },
+      ]);
+
+      const memoIdToRead = answers.memoToRead;
```
確認ですがこれはなぜ変数に代入しているのでしょうか？ また、`memoIdToRead` と `memoToRead` という名前だとぜんぜん違うものが入れ替えられるように見えます。

1mori replied:
該当の代入する処理は削除しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,136 @@
+      ]);
+
+      const memoIdToRead = answers.memoToRead;
+      const selectedMemo = rows.find((row) => row.id === memoIdToRead);
```
ID ではなくメモそのものを選択結果として取得できるようにすれば、この処理はいらなくなるんじゃないでしょうか。

1mori replied:
処理を変更し、削除しました。

1mori committed:
ignore: sqlite3をgitにアップロードしないように変更

1mori committed:
change: stdin, stdoutの置き換えを削除

1mori committed:
change: closeDatabaseをクラスの外で定義

1mori committed:
change: promise API -> callback API

1mori committed:
fix: err -> err.meassageでエラーメッセージのみ出力するように変更

1mori committed:
fix: 必要なエラーのみ捕捉する処理に変更

1mori committed:
change: テンプレート文字列に変更

1mori committed:
fix: ctrl+Cが押された時、メモが保存されないように修正

1mori committed:
change: userInput -> readlineInterface

1mori committed:
fix: runメソッド内のaddをasync化

1mori committed:
change: closeDatabaseの処理は各メソッド全てで実行されているので、処理を共通化

1mori committed:
change: closeDatabaseのエラー条件をより厳格に変更

1mori committed:
change: ブラケット記法からドット記法に変更

1mori committed:
change: rows -> memoRows

1mori committed:
fix: listメソッド内で必要なエラーを捕捉するように変更

1mori committed:
fix: tryで囲む範囲を限定、処理をメソッド化

1mori committed:
change: 省略記法を使用

1mori committed:
fix: 必要箇所のエラー捕捉に変更

1mori committed:
change: listメソッドのデータ取得処理もgetMemoRowsに統合

1mori committed:
change: データベースで取得するデータをワイルドカード記法に変更

1mori committed:
change: メモの入力処理を非同期化

1mori committed:
delete: 不必要な変数の宣言を削除

1mori committed:
change: メモのidは必要ないので、取得しないように変更

1mori committed:
define: chooseMemoメソッドを定義

1mori committed:
change: chooseMemoメソッドを使用するように変更

1mori committed:
change: deleteメソッドでもchooseMemoを使用するように変更

1mori committed:
delete: 使わなくなったresolveメソッドを削除

1mori committed:
change: deleteメソッドを考慮して、idも取得するように変更

1mori committed:
define: memo_utils.jsを定義、memo.js内のメソッドを移動

1mori committed:
add: 表示するメモが存在しない場合に終了する処理を追加

cafedomancer commented on 04.class/memo_utils.js:
```diff
@@ -0,0 +1,36 @@
+  try {
+    await closePromise(db);
+  } catch (err) {
+    if (err instanceof Error && err.code === "SQLITE_MISUSE")
+      console.error(err.message);
+    else throw err;
```
if 文は常にブロックを伴って書いてください。https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/if...else#%E8%A7%A3%E8%AA%AC

1mori replied:
修正しました。指摘ありがとうございます。

cafedomancer commented on 04.class/memo_utils.js:
```diff
@@ -0,0 +1,36 @@
+
+import { allPromise, closePromise } from "./db_utils.js";
+
+export async function closeDatabase(db) {
+  try {
+    await closePromise(db);
+  } catch (err) {
+    if (err instanceof Error && err.code === "SQLITE_MISUSE")
+      console.error(err.message);
+    else throw err;
+  }
+}
```
`memo_utils.js` というファイルと `closeDatabase(...)` という関数は名前的には何の関連もないように見えます。

1mori replied:
一度memo.jsに戻しました。
現状ではデータベースの操作関連をまとめた`memoDatabase`というクラスを作成し、そこに格納しています。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+    const readlineInterface = createInterface({ input: stdin, output: stdout });
+
+    await new Promise((resolve) => {
+      readlineInterface.on("line", (line) => {
+        inputText += `${line}\n`;
+      });
+
+      readlineInterface.on("close", resolve);
```
Promises API ではなく Callback API を使うことにした理由を教えてください。

1mori replied:
Promises APIには`question`という標準入力を受け付けるメソッドが存在しますが、`question`はEnterを押して改行した時点で入力受付を終了してしまいます。
今回のメモアプリは複数行の入力に対応させる必要があるため、Callback APIの`on`を使用しました。

Promises APIで`on`の使用例が掲載されていますが、callback APIと書き方が変わらないので今回はcallback APIを使用しました。

1mori replied:
質問です。
Promises APIには公式ドキュメント( https://nodejs.org/api/readline.html )を見る限り、他に複数行の入力に対応できるような処理は存在しないように見えます。
Promises APIにはそのような機能を持つメソッドがあるのですか？

cafedomancer replied:
ないです。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  async #add() {
+    let inputText = "";
+    const readlineInterface = createInterface({ input: stdin, output: stdout });
+
+    await new Promise((resolve) => {
+      readlineInterface.on("line", (line) => {
+        inputText += `${line}\n`;
+      });
+
+      readlineInterface.on("close", resolve);
+    });
```
readline の詳細を扱うのはメモの追加の本質的な処理ではないので、別のメソッドや関数に切り出したほうがよいと思います。

1mori replied:
`ReadlineInterface`クラスを定義し、そこから呼び出すようにしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+      await runPromise(this.db, "INSERT INTO memo (text) VALUES (?)", [
+        inputText,
+      ]);
+    } catch (err) {
+      if (err instanceof Error && err.code === "SQLITE_CONSTRAINT") {
+        console.error(err.message);
+      }
+    }
```
修正が不十分です。https://github.com/1mori/js-practices/pull/5#discussion_r1632487643

1mori replied:
```
else {
  throw err;
}
```

を後ろにつける形で修正しました。指摘ありがとうございます。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  async #list() {
+    const memoRows = await getMemoRows(this.db);
```
データベースのレコードであることを表すことがこの文脈で重要であるなら rows を付けたほうがいいと思いますが、そうでないのなら rows はいらないと思います。

1mori replied:
`getMemoRows`では`memo`テーブル内全てのデータを取得しています。そのため
`memoRows` -> `memos`
`memoRow` -> `memo`

に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+  async #list() {
+    const memoRows = await getMemoRows(this.db);
+
+    memoRows.forEach((row) => {
```
`row` に関して、修正が不十分です。https://github.com/1mori/js-practices/pull/5#discussion_r1632490076

1mori replied:
修正しました。指摘ありがとうございます。
二箇所存在しており、それぞれ`row` -> `memoRow`に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+      console.log("表示するメモがありません。");
+      return;
+    }
+    const answers = await chooseMemo(
```
格納される値が配列なら複数形にしたほうがいいと思いますが、そうでないのなら複数形にするのは誤解を招きそうです。

1mori replied:
おっしゃる通り、ここは一つのメモが選択されているので`answer`にしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,106 @@
+      "Choose a memo you want to see:",
+    );
+
+    const selectedMemo = answers.memoToRead.text;
```
代入されているのはメモのオブジェクトではなくメモの内容を表すテキストなので、それに memo と付けてしまうのはどうかなーと思います。

1mori replied:
`selectedText`に変更しました。指摘ありがとうございます。

cafedomancer commented on 04.class/memo_utils.js:
```diff
@@ -0,0 +1,36 @@
+  return memoRows;
+}
+
+export async function chooseMemo(memoRows, name, type, message) {
```
`chooseMemo` という名前から考えるに、これはメモアプリの本質的な処理じゃないですかね。`MemoApp` クラスから切り出されている理由がよく分からないです。

1mori replied:
MemoApp内のプライベートメソッドとして定義しました。

1mori committed:
fix: ブロック記法がされていないif文を修正

1mori committed:
change: class dbPromiseを定義、データベースの非同期処理系を一つのクラスに格納

1mori committed:
remove: memo_utils.jsを削除、中身の関数を一度memo.jsに移動

1mori committed:
fix: 想定外のエラーを握り潰していたため、エラー処理の修正

1mori committed:
fix: row -> memoRow

1mori committed:
change: selectedMemo -> selectedText

1mori committed:
change: answers -> answer

1mori committed:
change: データベース全体のデータを取得し、格納する変数の名前をmemoRows -> memosに変更

1mori committed:
define: memoDataBAseを定義、データベースの読み込みと存在確認を実行する

1mori committed:
fix: chooseMemoをMemoApp内プライベート関数として定義

1mori committed:
add: 改行追加

1mori committed:
データベース関連操作のため、getMemoRows, closeをメモDataBaseに追加

1mori committed:
change: readline系の処理をMemoApp内から除外、外に出す

1mori committed:
change: readline処理をクラス化

1mori committed:
define: memoDatabase内にinsertメソッドを定義

1mori committed:
change: 名称変更　getMemoRows -> all

1mori committed:
change: chooseMemo -> choose

1mori committed:
define: memoDatabase内にdeleteメソッドを追加

1mori committed:
fix: closeメソッド内のdbの呼び出しをthis.dbに修正

1mori committed:
change: テーブルの初期作成メソッドをensureTableExistsに変更

1mori committed:
change: memo -> memos(sql)

1mori committed:
change: MemoApp実行の行を一行に短縮

1mori committed:
fix: db_utils.jsの呼び出しをクラス外でしていたので修正

1mori committed:
fix: const appを削除

1mori committed:
delete: 標準モジュールの明示的な宣言を削除

1mori committed:
fix: クラス名の大文字の修正

1mori committed:
fix: readlineのimportの消去を戻した

1mori committed:
change: memoRows -> memos

cafedomancer commented on 04.class/db_utils.js:
```diff
@@ -0,0 +1,39 @@
+      });
+    });
+  }
+}
```
状態 (プロパティ) を持たないクラスは定義しても意味がないです。ただ関数がまとまっているだけだからです。ただし、これらの処理をクラスにする必要がないとは言っていない点に注意してください (必要があるとも言っていない) 。現状のコードに対してクラスにする意味がないと言っているだけです。

1mori replied:
`db_utils.js`内で定義していたクラスは削除し、`MemoDatabase`内メソッドに処理を追加するようにしました。

cafedomancer commented on 04.class/db_utils.js:
```diff
@@ -0,0 +1,39 @@
+class DbPromise {
```
メモアプリにおいては node-sqlite3 の関数を promise 化することが重要なわけではないので、この名前は適当ではないと思います。

1mori replied:
指摘ありがとうございます。今回はクラスを削除したので、次回に活かします。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env node
+
+import { createInterface } from "readline";
```
`createInterface` という名前で import するとプログラム中で readline の関数であることが分からなくなりそうです。

1mori replied:
`import readline from "readline";`とし、`readline.createInterface`で呼び出すように変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+      });
+    });
+  }
+}
```
プログラム中でクラスが 1 つではなくなったのならクラスはクラスごとに別のファイルに定義してください。

1mori replied:
定義していた`MemoDatabase`, `ReadlineInterface`は別ファイルに移動しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+  }
+}
+
+class memoDatabase {
```
クラス名は大文字始まりの camel case にしてください。

1mori replied:
修正させていただきました。
このクラスは別ファイルにて定義させていただきます。

cafedomancer commented on 04.class/db_utils.js:
ファイル名とクラス名が一致していません。

1mori replied:
指摘ありがとうございます。次回以降に注意します。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+    return new Promise((resolve) => {
+      let text = "";
+      this.readlineInterface.on("line", (line) => {
+        text += `${line}\n`;
```
見落としていましたが、この実装だと入力に加えて最後に余分な改行が追加されてしまうと思います。

1mori replied:
最終文字が`\n`の場合に`.slice`を用いて削除する処理を追加しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+    this.memoDatabase.delete(id);
+  }
+
+  async #choose(memos, name, type, message) {
```
こんなに引数は必要ないような。本当に可変な部分だけ引数で受け取るようにしてください。

1mori replied:
`name`はmemoで統一しました。
`type`は今回listしか使用していないため、直接指定するようにしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+      await this.#add();
+    }
+    await this.memoDatabase.close(this.db);
+  }
```
見通していましたが、public メソッドは private メソッドよりも前に定義してください。

1mori replied:
今回は`run()`が該当していたので、プライベートメソッドの上に移動させました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+class memoDatabase {
+  constructor() {
+    this.db = new sqlite3.Database("./memo.sqlite3");
+    this.promise = new DbPromise();
```
`this.promise` だけでは格納されているものが正しく表現できていないと思います。この名前だと Promise オブジェクトが格納されているのかと勘違いしてしまいます。

1mori replied:
指摘ありがとうございます。今回はDbPromiseを削除したので次回以降注意します。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+  async all() {
+    let memos;
+    try {
+      memos = await this.promise.all(this.db, "SELECT id, text FROM memos");
```
このクエリだとメモの順序が不定になりそうです。

1mori replied:
`ORDER BY id`で順序を固定しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+    return memos;
+  }
+
+  async insert(input) {
```
このメソッドに関してプロンプトが絡んでいるなら input だと思いますが、この文脈ではそうではないので名前が適当ではないと思います。

1mori replied:
`memo`に変更しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,190 @@
+
+      this.readlineInterface.on("close", () => {
+        resolve(text);
+      });
```
この実装だと入力途中に Ctrl-C でプログラムを停止しても処理が続行されてしまいそうです。

1mori replied:
`Ctrl+C`を検知するイベント`SIGINT`を利用し、強制終了した際にはエラーを発生させて保存しないように変更しました。
https://github.com/1mori/js-practices/pull/5/commits/d3c6b629c5d2c3efe75351b96e39803dcffb3d25

最初は`process.exit()`で終了させていましたが、データベースが正常に閉じられないのでエラーを利用しました。

1mori committed:
fix: memoDatabase -> MemoDatabase

1mori committed:
define: memo-databaseファイルを作成、MemoDatabaseを移行

1mori committed:
define: readline-interface.jsを作成、import readline from 'readline'に変更することでよりわかりやすく

1mori committed:
change: run()はプライベートメソッドではないので、パブリックメソッドよりも上に移動

1mori committed:
fix: 最終行の一行多い改行を削除できるように修正

最後の文字が改行文字であった場合に、`.slice`で改行文字を除去する

1mori committed:
all()でのクエリの順序をorder byで固定化

1mori committed:
change: textをクラス内変数化

1mori committed:
change: input -> memo

1mori committed:
add: 標準入力中にctrl+Cを検知した場合、プログラムが終了するように変更

Ctrl+Cを監視するイベント"SIGINT"を設定し、検知した場合は`process.exit();`で終了するようにした

参考
https://nodejs.org/api/readline.html#event-sigint

1mori committed:
add: 強制終了時にデータベースが正常に閉じられるようにエラーハンドリングを設定

SIGINTイベントを検知した際、`process.exit();`とするのではなく、
エラーを発生させてデータベースへの保存を回避し、データベースを閉じる処理を完了させるようにした

1mori committed:
change: 同一のオプションを指定していたため、inquirerで指定する引数を減らした

1mori committed:
change: `db_utils.js`の処理を`memo-database.js`に組み込み削除

db_utils.jsでは非同期処理を設定していたが、MemoDatabaseからさらに呼び出すようになっていたので簡単化のために処理を統合
また、値を保持しないクラスは定義の意味がないため削除

1mori committed:
fix: エラー処理の修正

1mori committed:
lint: ツールのインデント最終確認

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+import MemoDatabase from "./memo-database.js";
+import ReadlineInterface from "./readline-interface.js";
+
+class MemoApp {
```
修正が漏れています 😵 https://github.com/1mori/js-practices/pull/5#discussion_r1638175425

1mori replied:
現在のファイルを見る限り、現在のコードは定義したクラス・作成したファイル共に3つになっており、それぞれのファイルごとにクラスが1つずつ定義されている状態です。
引用しているリンク先のコメントの指摘の内容を見る限り、特におかしいところは見られないです。

何を直したらいいのかいいいのかわからないので、もう一度確認していただいてもよろしいでしょうか。🙏

1mori replied:
指摘範囲からして、リンクはこちらですか？https://github.com/1mori/js-practices/pull/5#discussion_r1638176769
camelCaseに変更しました。

cafedomancer replied:
別途コメントしました。

cafedomancer commented on 04.class/memo-database.js:
```diff
@@ -0,0 +1,74 @@
+          resolve(this);
+        }
+      });
+    });
```
run や all を proimse でラップしている部分がクラス内で重複しています。

1mori replied:
エラー処理をクラス内プライベートメソッド`#handleError`にまとめました。

cafedomancer replied:
別途コメントしました。

cafedomancer commented on 04.class/memo-database.js:
```diff
@@ -0,0 +1,74 @@
+    this.db = new sqlite3.Database("./memo.sqlite3");
+  }
+
+  async ensureTableExists() {
```
メソッド内で `await` を使っていないのに `async` が付いています。

1mori replied:
削除しました。指摘ありがとうございます。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+    } else {
+      await this.#add();
+    }
+    await this.memoDatabase.close(this.db);
```
`close()` は引数を取らないメソッドです 😵

1mori replied:
削除しました。ありがとうございます。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+
+    const answer = await this.#choose(memos, "Choose a memo you want to see:");
+
+    const selectedText = answer.memo.text;
```
select されたのは text ではなく memo のような気がしました。77 行目の `id` と名前に一貫性がない点も気になります。

1mori replied:
変数が必要なくなったために削除し、直接表示するようにしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+    const answer = await this.#choose(memos, "Choose a memo you want to see:");
+
+    const selectedText = answer.memo.text;
+    if (selectedText) {
```
`selectedText` がないケースがよく分からなかったのですが、どんなケースでしょうか？ 🤔

1mori replied:
当初はinquirerの挙動がわからず、過剰に確認を設けていました。削除しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+      "Choose a memo you want to delete:",
+    );
+
+    const id = answer.memo.id;
```
変数に入れなくてもいいかなと思います。

1mori replied:
削除しました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  async #choose(memos, message) {
+    const choices = memos.map((memoRow) => ({
```
`memos` に対応させるなら `memoRow` ではなく `memo` かなーと思います。

1mori replied:
変更しました。

cafedomancer commented on 04.class/memo-database.js:
```diff
@@ -0,0 +1,74 @@
+    });
+  }
+
+  async insert(memo) {
```
受け取るのはメモではなくメモのテキストだと思います 👀

1mori replied:
変更しました。

cafedomancer commented on 04.class/memo-database.js:
```diff
@@ -0,0 +1,74 @@
+    });
+  }
+
+  async all() {
```
メソッド名は動詞始まりにしてください。

1mori replied:
`fetchAll`に変更しました。

cafedomancer commented on 04.class/readline-interface.js:
```diff
@@ -0,0 +1,39 @@
+    this.text = "";
+    this.sigintReceived = false;
+
+    this.readlineInterface.on("SIGINT", () => {
+      this.sigintReceived = true;
+      this.readlineInterface.close();
+    });
```
この処理が `inputText()` の中ではなく `constructor()` の中にある理由が分からなかったのですが、教えてもらってもよいでしょうか？

1mori replied:
promiseの中で宣言すると正常に動作するのかわからなかったので、予め宣言する形で実装しました。
問題なく動作することがわかったので、現在は他の`.on`メソッドと同様にpromise内で宣言するように変更しました。

cafedomancer commented on 04.class/readline-interface.js:
```diff
@@ -0,0 +1,39 @@
+        if (this.sigintReceived) {
+          reject(new Error("SIGINT received"));
+        } else {
+          if (this.text.endsWith("\n")) {
+            this.text = this.text.slice(0, -1);
+          }
```
改行を削るような処理を追加するのではなく、そもそも余分な改行を追加しないような処理に修正したほうがよいのではないかなーと思います。

1mori replied:
配列で要素を1行ずつ受け取り、最後に`\n`で結合するようにしました。

1mori committed:
delete: 不要なasyncの削除

1mori committed:
fix: 引数を取らないcloseメソッドに指定していた引数の削除

1mori committed:
change: 選択されたメモの挙動を修正

1mori committed:
change: memoRow -> memo

1mori committed:
change: memo -> text

1mori committed:
change: all -> fetchAll

1mori committed:
chenge: echoでメモを入力した場合、メモが再度表示されないように変更

createInterfaceを作る際、outputにprocess.stdoutと指定していたのを、nullに変更

1mori committed:
fix: 強制終了処理をpromise内で宣言するように修正した

1mori committed:
change: 標準入力された文字を配列で受け取るように変更した

1mori committed:
change: エラー処理を一つにまとめた

エラー処理が重複していたため、handleErrorにまとめた

1mori committed:
change: ファイル名とクラス名の結びつけ

1mori committed:
lint: npm run fix

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,91 @@
+import MemoDatabase from "./memoDatabase.js";
+import ReadlineInterface from "./readlineInterface.js";
+
+class Memo {
```
クラスはクラス定義のみを含むファイルに別途定義してください :pray: 他のクラスと定義方法が一貫していないからというのと、もしこのクラスがこのファイルで export されていた場合、トップレベルに定義している他の内容が import 時に実行されてしまうから、というのが理由です。

1mori replied:
新たに`memoApp.js`を定義し、Memoクラス（改名後はMemoAppクラス)をそちらに移動させました。
従来の`memo.js`ではMemoAppを呼び出して実行するだけのコードに変更しました。

cafedomancer commented on 04.class/memoDatabase.js:
```diff
@@ -0,0 +1,52 @@
+  ensureTableExists() {
+    const query =
+      "CREATE TABLE IF NOT EXISTS memos (id INTEGER PRIMARY KEY AUTOINCREMENT, text TEXT NOT NULL)";
+    return new Promise((resolve) => {
+      this.db.run(query, (err) => this.#handleError(err, resolve, this));
+    });
```
https://github.com/1mori/js-practices/pull/5#discussion_r1645841298 に対する修正方針が合っていなさそうです 😵 エラーハンドリングについてではなく、コールバック形式の処理を Promise でラップする部分についてコメントしています。

1mori replied:
Promiseでの非同期処理をプライベートメソッド化しました。

cafedomancer commented on 04.class/readlineInterface.js:
```diff
@@ -0,0 +1,36 @@
+  constructor() {
+    this.readlineInterface = readline.createInterface({
+      input: process.stdin,
+      output: null,
```
`output` の指定を省略すれば `undefined` になるはずなので、わざわざ `null` を指定する必要はないかなーと思いました。

1mori replied:
削除しました。指摘ありがとうございます。

cafedomancer commented on 04.class/readlineInterface.js:
```diff
@@ -0,0 +1,36 @@
+    });
+
+    this.lines = [];
+    this.sigintReceived = false;
```
`SIGINT` イベントを捕捉した時点で promise を reject すればいいので、この変数は不要になっている気がしました。

1mori replied:
ありがとうございます。変更しました。

1mori committed:
delete: 標準出力のoutputの項目を削除

1mori committed:
change: git mv memo.js memoApp.js

1mori committed:
change: 強制終了時に直接rejectを返すようにした

1mori committed:
change: クラスのファイル呼び出し定義を変更

新たにmemoApp.jsを作成、memo.jsからMemoAppクラスを呼び出す形にした

1mori committed:
change: promiseそのものをメソッド化
